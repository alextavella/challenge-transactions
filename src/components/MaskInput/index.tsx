import React, { useCallback, useMemo } from 'react';
import { ValidationRules } from 'react-hook-form';
import {
  CPF,
  CREDIT_CARD_CVV,
  CREDIT_CARD_EXPIRATION_DATE,
  CREDIT_CARD_NUMBER,
  CURRENCY,
  MaskRule,
} from '../../utils/patterns';
import Input, { InputProps } from '../Input';

export enum MaskInputType {
  buyer_document = 'buyer_document',
  credit_card_number = 'credit_card_number',
  credit_card_expiration_date = 'credit_card_expiration_date',
  credit_card_cvv = 'credit_card_cvv',
  currency = 'currency',
}

export interface MaskInputProps extends InputProps {
  mask: MaskInputType;
  rules?: ValidationRules;
  onKeyUp?: (event: any) => void;
}

interface MaskRuleType {
  [key: string]: MaskRule;
}

const maskRules: MaskRuleType = {
  [MaskInputType.buyer_document]: CPF,
  [MaskInputType.credit_card_number]: CREDIT_CARD_NUMBER,
  [MaskInputType.credit_card_expiration_date]: CREDIT_CARD_EXPIRATION_DATE,
  [MaskInputType.credit_card_cvv]: CREDIT_CARD_CVV,
  [MaskInputType.currency]: CURRENCY,
};

type RefProps = (rules: ValidationRules) => any;

const MaskInput: React.ForwardRefRenderFunction<RefProps, MaskInputProps> = (
  { name, label, mask, type = 'tel', rules = {}, onKeyUp, ...rest },
  innerRef: any,
) => {
  const maskRule = useMemo<MaskRule>(() => maskRules[mask], [mask]);

  const handleChange = useCallback(
    (event: any) => {
      event.target.value = maskRule.transform(event.target.value);
      onKeyUp && onKeyUp(event);
    },
    [maskRule, onKeyUp],
  );

  return (
    <Input
      {...rest}
      type={type}
      name={name}
      label={label}
      onKeyUp={handleChange}
      ref={
        innerRef &&
        innerRef({
          ...rules,
          ...maskRule.rules,
        })
      }
    />
  );
};

export default React.forwardRef(MaskInput);
