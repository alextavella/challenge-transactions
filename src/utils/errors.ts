import { ValidationError } from 'yup';

interface Error {
  [key: string]: string;
}

export function getValidationErrors(errors: ValidationError): any {
  return errors.inner.reduce((acc: Error, error: ValidationError) => {
    return { ...acc, [error.path]: error.message };
  }, {});
}
