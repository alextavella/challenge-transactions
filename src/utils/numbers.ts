export const formatCurrency = (value: number): string => {
  const formatted = new Intl.NumberFormat([], {
    style: 'currency',
    currency: 'BRL',
    minimumFractionDigits: 2,
  }).format(value);

  return formatted
    .replace('.', '#')
    .replace(/,/g, '.')
    .replace('#', ',')
    .replace('R$', 'R$ ');
};

export const unformatCurrency = (value: string): number => {
  return Number(
    value
      .replace('R$ ', '')
      .replace(',', '#')
      .replace(/\./g, '')
      .replace('#', '.'),
  );
};
