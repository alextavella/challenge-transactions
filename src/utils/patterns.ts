import { ValidationRules } from 'react-hook-form';

export interface MaskRule {
  rules: ValidationRules;
  transform: (text: string) => string;
}

export const CPF: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/^([0-9]{3}).([0-9]{3}).([0-9]{3})-([0-9]{2})$/),
      message: 'CPF inválido',
    },
  },
  transform: (text: string = ''): string => {
    const result = text
      .replace(/\D/g, '')
      .match(/(\d{0,3})(\d{0,3})(\d{0,3})(\d{0,2})/)!;

    const part1 = result[2] ? `${result[1]}.${result[2]}` : result[1];
    const part2 = result[3] ? `.${result[3]}` : '';
    const part3 = result[4] ? `-${result[4]}` : '';

    return `${part1}${part2}${part3}`;
  },
};

export const CREDIT_CARD_NUMBER: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/^([0-9]{4})\s([0-9]{4})\s([0-9]{4})\s([0-9]{4})$/),
      message: 'Número do cartão inválido',
    },
  },
  transform: (text: string = ''): string => {
    const result = text
      .replace(/\D/g, '')
      .match(/(\d{0,4})(\d{0,4})(\d{0,4})(\d{0,4})/)!;

    const part1 = result[2] ? `${result[1]} ${result[2]}` : result[1];
    const part2 = result[3] ? ` ${result[3]}` : '';
    const part3 = result[4] ? ` ${result[4]}` : '';

    return `${part1}${part2}${part3}`;
  },
};

export const CREDIT_CARD_EXPIRATION_DATE: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/^(0[1-9]|1[012])\/?(20[2-9][0-9])$/),
      message: 'Data de expiração inválida',
    },
  },
  transform: (text: string = ''): string => {
    const result = text.replace(/\D/g, '').match(/(\d{0,2})(\d{0,4})/)!;

    return result[2] ? `${result[1]}/${result[2]}` : result[1];
  },
};

export const CREDIT_CARD_CVV: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/[0-9]{3}/),
      message: 'CVV inválido',
    },
  },
  transform: (text: string = ''): string => {
    return text.substr(0, 3);
  },
};

export const CURRENCY: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/([0-9]\.?)?[0-9]+,[0-9]{2}/),
      message: 'Valor inválido',
    },
  },
  transform: (text: string = ''): string => {
    let number = text.replace(/[^0-9]/g, '');

    if (number.length === 0) number = '0.00';
    else if (number.length === 1) number = '0.0' + number;
    else if (number.length === 2) number = '0.' + number;
    else
      number =
        number.substring(0, number.length - 2) +
        '.' +
        number.substring(number.length - 2, number.length);

    number = Number(number).toFixed(2);
    number = number.replace(/\./g, ',');

    const x = number.split(',');
    const x1 = x[0];
    const x2 = x.length > 1 ? ',' + x[1] : '';

    const formatAmountNoDecimals = (number: string) => {
      var rgx = /(\d+)(\d{3})/;
      while (rgx.test(number)) {
        number = number.replace(rgx, `$1.$2`);
      }
      return number;
    };

    return `R$ ` + formatAmountNoDecimals(x1) + x2;
  },
};
