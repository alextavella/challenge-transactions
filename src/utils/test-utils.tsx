import { render } from '@testing-library/react';
import React from 'react';
import MainProvider from '../providers';

const AllTheProviders: React.FC = ({ children }) => {
  return <MainProvider>{children}</MainProvider>;
};

const customRender = (ui: React.ReactElement, options?: any) =>
  render(ui, {
    wrapper: AllTheProviders,
    ...options,
  });

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
