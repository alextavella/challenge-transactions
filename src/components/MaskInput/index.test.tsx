import React from 'react';
import MaskInput, { MaskInputType } from '.';
import {
  CPF,
  CREDIT_CARD_CVV,
  CREDIT_CARD_EXPIRATION_DATE,
  CREDIT_CARD_NUMBER,
  CURRENCY,
} from '../../utils/patterns';
import { cleanup, fireEvent, render, screen } from '../../utils/test-utils';

describe('<MaskInput />', () => {
  afterEach(() => {
    cleanup();
  });

  it('should be able to render correctly', () => {
    // Arrange
    const spyRegister = jest.fn();

    const { baseElement } = render(
      <MaskInput
        name="cpf"
        label="CPF"
        mask={MaskInputType.buyer_document}
        ref={spyRegister}
      />,
    );

    // Assert
    expect(baseElement).toMatchSnapshot();
  });

  it('should be able to call function when change', () => {
    // Arrange
    const spyKeyUp = jest.fn();

    render(
      <MaskInput
        name="cpf"
        label="CPF"
        mask={MaskInputType.buyer_document}
        ref={jest.fn()}
        onKeyUp={spyKeyUp}
      />,
    );

    const input = screen.getByTestId('cpf') as HTMLInputElement;

    // Act
    fireEvent.keyUp(input, { target: { value: '1234' } });

    // Assert
    expect(spyKeyUp).toHaveBeenCalled();
    expect(input.value).toBe('123.4');
  });

  it('should be able to validation cpf when change', () => {
    // Arrange
    const spyRegister = jest.fn();
    const rules = {
      required: 'Required',
      ...CPF.rules,
    };

    render(
      <MaskInput
        name="cpf"
        label="CPF"
        mask={MaskInputType.buyer_document}
        ref={spyRegister}
        rules={rules}
      />,
    );

    const input = screen.getByTestId('cpf') as HTMLInputElement;

    // Act
    fireEvent.change(input, { target: { value: '123' } });

    // Assert
    expect(spyRegister).toHaveBeenCalledWith(rules);
  });

  it('should be able to validation cvv when change', () => {
    // Arrange
    const spyRegister = jest.fn();
    const rules = {
      required: 'Required',
      ...CREDIT_CARD_CVV.rules,
    };

    render(
      <MaskInput
        name="cvv"
        label="CVV"
        mask={MaskInputType.credit_card_cvv}
        ref={spyRegister}
        rules={rules}
      />,
    );

    const input = screen.getByTestId('cvv') as HTMLInputElement;

    // Act
    fireEvent.change(input, { target: { value: '123' } });

    // Assert
    expect(spyRegister).toHaveBeenCalledWith(rules);
  });

  it('should be able to validation creditcard number when change', () => {
    // Arrange
    const spyRegister = jest.fn();
    const rules = {
      required: 'Required',
      ...CREDIT_CARD_NUMBER.rules,
    };

    render(
      <MaskInput
        name="creditcard_number"
        label="Creditcard number"
        mask={MaskInputType.credit_card_number}
        ref={spyRegister}
        rules={rules}
      />,
    );

    const input = screen.getByTestId('creditcard_number') as HTMLInputElement;

    // Act
    fireEvent.change(input, { target: { value: '1234' } });

    // Assert
    expect(spyRegister).toHaveBeenCalledWith(rules);
  });

  it('should be able to validation creditcard expiration date when change', () => {
    // Arrange
    const spyRegister = jest.fn();
    const rules = {
      required: 'Required',
      ...CREDIT_CARD_EXPIRATION_DATE.rules,
    };

    render(
      <MaskInput
        name="creditcard_expiration_date"
        label="Creditcard number"
        mask={MaskInputType.credit_card_expiration_date}
        ref={spyRegister}
        rules={rules}
      />,
    );

    const input = screen.getByTestId(
      'creditcard_expiration_date',
    ) as HTMLInputElement;

    // Act
    fireEvent.change(input, { target: { value: '1234' } });

    // Assert
    expect(spyRegister).toHaveBeenCalledWith(rules);
  });

  it('should be able to validation currency when change', () => {
    // Arrange
    const spyRegister = jest.fn();
    const rules = {
      required: 'Required',
      ...CURRENCY.rules,
    };

    render(
      <MaskInput
        name="currency"
        label="Currency"
        mask={MaskInputType.currency}
        ref={spyRegister}
        rules={rules}
      />,
    );

    const input = screen.getByTestId('currency') as HTMLInputElement;

    // Act
    fireEvent.change(input, { target: { value: '1234' } });

    // Assert
    expect(spyRegister).toHaveBeenCalledWith(rules);
  });

  // describe('buyer_document', () => {
  // it('should be able to render correctly when buyer_document', () => {
  //   // Arrange
  //   const spyRegister = jest.fn();

  //   render(
  //     <MaskInput
  //       name="cpf"
  //       label="CPF"
  //       mask={MaskInputType.buyer_document}
  //       ref={spyRegister}
  //     />,
  //   );

  //   const input = screen.getByTestId('cpf');

  //   // Act
  //   fireEvent.change(input, {
  //     target: { value: '11122233344' },
  //   });

  //   // Assert
  //   expect(screen.queryByDisplayValue('111.222.333-44')).toBeTruthy();
  // });
  // });

  // describe('credit_card_cvv', () => {
  //   it('should be able to render correctly', () => {
  //     // Arrange
  //     render(
  //       <MaskInput
  //         name="cvv"
  //         label="CVV"
  //         mask={MaskInputType.credit_card_cvv}
  //       />,
  //     );

  //     const input = screen.getByTestId('cvv');

  //     // Act
  //     fireEvent.change(input, {
  //       target: { value: '123' },
  //     });

  //     // Assert
  //     expect(screen.queryByDisplayValue('123')).toBeTruthy();
  //   });
  // });

  // describe('credit_card_expiration_date', () => {
  //   it('should be able to render correctly', () => {
  //     // Arrange
  //     render(
  //       <MaskInput
  //         name="expirate"
  //         label="Expirate date"
  //         mask={MaskInputType.credit_card_expiration_date}
  //       />,
  //     );

  //     const input = screen.getByTestId('expirate');

  //     // Act
  //     fireEvent.change(input, {
  //       target: { value: '122020' },
  //     });

  //     // Assert
  //     expect(screen.queryByDisplayValue('12/2020')).toBeTruthy();
  //   });
  // });

  // describe('credit_card_number', () => {
  //   it('should be able to render correctly', () => {
  //     // Arrange
  //     render(
  //       <MaskInput
  //         name="number"
  //         label="Card number"
  //         mask={MaskInputType.credit_card_number}
  //       />,
  //     );

  //     const input = screen.getByTestId('number');

  //     // Act
  //     fireEvent.change(input, {
  //       target: { value: '1111222233334444' },
  //     });

  //     // Assert
  //     expect(screen.queryByDisplayValue('1111 2222 3333 4444')).toBeTruthy();
  //   });
  // });

  // describe('currency', () => {
  //   it('should be able to render correctly', () => {
  //     // Arrange
  //     render(
  //       <MaskInput
  //         name="amount"
  //         label="Currency"
  //         mask={MaskInputType.currency}
  //       />,
  //     );

  //     const input = screen.getByTestId('amount');

  //     // Act
  //     fireEvent.change(input, {
  //       target: { value: '1234' },
  //     });

  //     // Assert
  //     expect(screen.queryByDisplayValue('R$ 12,34')).toBeTruthy();
  //   });
  // });
});
