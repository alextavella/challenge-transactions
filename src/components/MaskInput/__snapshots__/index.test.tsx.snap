// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MaskInput /> should be able to render correctly 1`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink="false"
      >
        CPF
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          data-testid="cpf"
          name="cpf"
          type="tel"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-3"
          >
            <span>
              CPF
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</body>
`;
